wasp/keyboard.py

```diff
@@ -0,0 +1,272 @@
+                    self._string = self._string[:-1] + value[self._count]
+                else:
+                    self._last = value
+                    self._count = 0
+                    self._string += value[self._count]
+                self._last_time = time.time()
+                self._display_string()
+            else:
+                if value == "bdel":
+                    self._string = self._string[:-1]
+                    self._display_string()
+                elif value == "fdel":
+                    self._string_after = self._string_after[1:]
+                    self._display_string()
+                elif value == "barr":
+                    self._string_after = self._string[-1] + self._string_after
+                    self._string = self._string[:-1]
+                    self._display_string()
+                elif value == "farr":
+                    self._string = self._string + self._string_after[:1]
+                    self._string_after = self._string_after[1:]
+                    self._display_string()
+                elif value == "shift":
+                    self._cap_mode = (self._cap_mode + 1) % 3
+                    self._draw()
+                    self._last = ""
+                    self._count = None
+                    self._last_time = 0
+                elif value == "sym":
+                    self._cap_mode = 0
+                    self._screen = (self._screen + 1) % len(display)
+                    self._draw()
+
+    def _display_string(self):
+        draw = wasp.watch.drawable
+        theme = wasp.system.theme
+        hi = theme("bright")
+        draw.set_color(hi)
+
+        # TODO: make sure cursor is on screen
+        string = f"{self._string}|{self._string_after}"
+        boundaries = draw.wrap("".join(reversed(string)), 194, False)
+        draw.string(
+            string[0 - boundaries[1] :],
+            0,
+            10,
+            width=196,
+            right=True,
+        )
+
+    def _draw_grid(self, draw, top_bar_bg, grid_bg, line_color):
+        # Draw the background
+        draw.fill(top_bar_bg, 0, 0, 240, 48)
+        draw.fill(grid_bg, 0, 48, 240, 236 - 48)
+
+        # Make grid:
+        draw.set_color(line_color)
+        for i in range(4):
+            # horizontal lines
+            draw.line(x0=0, y0=(i + 1) * 47, x1=239, y1=(i + 1) * 47)
+            # vertical lines
+            draw.line(x0=(i + 1) * 47, y0=47, x1=(i + 1) * 47, y1=235)
+        draw.line(x0=0, y0=47, x1=0, y1=236)
+        draw.line(x0=239, y0=47, x1=239, y1=236)
+        draw.line(x0=0, y0=236, x1=239, y1=236)
+
+    def _draw(self):
+        draw = wasp.watch.drawable
+        theme = wasp.system.theme
+
+        hi = theme("bright")
+        lo = theme("mid")
+        mid = draw.lighten(lo, 2)
+        bg = draw.darken(theme("ui"), theme("contrast"))
+        bg2 = draw.darken(bg, 2)
+
+        self._draw_grid(draw, 0, bg2, lo)
```

Comment:
If this line is moved to the end of the `open` function, just before the `self._draw()` call, then redraws for the shift or symbols keys will be somewhat faster.